flink-autoscaler/src/main/java/org/apache/flink/autoscaler/event/AutoscalerEventUtils.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.autoscaler.event;
+
+import org.apache.flink.annotation.Experimental;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/** The utils of {@link AutoScalerEventHandler}. */
+@Experimental
+public class AutoscalerEventUtils {
+
+    private static final Pattern SCALING_REPORT_SEPARATOR = Pattern.compile("\\{(.+?)\\}");
+    private static final Pattern VERTEX_SCALING_REPORT_PATTERN =
+            Pattern.compile(
+                    "Vertex ID (.*?) \\| Parallelism (.*?) -> (.*?) \\| Processing capacity (.*?) -> (.*?) \\| Target data rate (.*)");
+
+    /** Parse the scaling report from original scaling report event. */
+    public static List<VertexScalingReport> parseVertexScalingReports(String scalingReport) {
```

Comment:
Why are we parsing the output here? Just for the tests? I would move this class to the test namespace.

Reply:
Keep it here for 2 reasons:

1. For test and ensure the ScalingReport can be parsed correctly
    - In the future, we need to be careful when modifying the contents of `ScalingReport` because its format is being parsed by some users.
2. The public utils can be used for all autoscaler users.
    - That's why I added the `@Experimental` for `AutoscalerEventUtils` class.

For reason1, keep it here or test namespace are fine. For reason2, keep it here is easy to use for users.

I can move it to test namespace if you think it's not necessary, thank you

Reply:
I'm not 100% sure whether we should provide utils for parsing the events. Users have the spec, metrics, and the reconciliation status to programmatically extract the deployment state. Deployment events are not meant to be machine-readable. Maybe we are missing important infos in the status field? I would rather add scaling information there instead of parsing the events.

Reply:
Thank you @mxm , I will do more research next week to check how to let other services to get the recommend parallelism properly.

Reply:
Thank you. If it is common practice to parse deployment events and there is no other good way to get this data, I'm also open to perusing the current approach.

Reply:
>  Users have the spec, metrics, and the reconciliation status to programmatically extract the deployment state

`Spec` and `Reconciliation status` are related to kubernetes, right?

After analysis, I think the metric report isn't suitable here. In general, the metric is reported or fetched periodically,
- When it's too frequent, the reported recommend parallelism is always same (It isn't needed)
- When it's too infrequent, we might miss some recommend parallelism.

So, I think the event driven is suitable for this requirement, and the event handler is the event driven. Also, the core change of this PR only added `{` and `}` between multiple vertices. If you think we don't need to provide the parse utils class, I can move it to the test namespace.

What's your opinion?

Reply:
Makes sense. I was thinking to expose the overrides on the status field of the custom resource, but that would require pull instead of push. So no objections from my side moving ahead with the proposed changes.

Reply:
Yeah, it needs push model instead pull model.

> So no objections from my side moving ahead with the proposed changes.

This prod code and test code is already ready.